subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java

```diff
@@ -71,17 +71,22 @@ public String toString() {
     }
 
     @Override
-    public void create(final Action<? super T> action) {
+    public void create(Action<? super T> action) {
+        create(getElementTypeClass(), action);
+    }
+
+    @Override
+    public <S extends T> void create(Class<S> type, final Action<? super T> action) {
         state.assertCanMutate();
 
-        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(elementType)));
+        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(type)));
```

Comment:
This won't work when mixing different types in a `ModelSet` as node paths will overlap.
Simply reverting the change on that line fix the issue.